File: ros2_ws/src/s5_cpp_camera/src/camera.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <image_transport/image_transport.hpp>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include <chrono>

void find_devices(){
    for (int i = 0; i < 5; ++i){ // Check up to 5 devices
        cv::VideoCapture cap_test(i);
        if (cap_test.isOpened()) {
            std::cout << "Device " << i << " is available." << std::endl;
        }
        // Release the VideoCapture object
        cap_test.release();
    }
}

class CameraNode : public rclcpp::Node {
    public:
        CameraNode(): 
            Node("cpp_camera_node"), frame_count_(0), fps_(0.0) {
                RCLCPP_INFO(this->get_logger(), "C++ Camera node has been started");

                // Check available camera IDs
                // find_devices();

                // Declare parameters with default values
                this->declare_parameter<int>("camera.deviceID", 0);
                this->declare_parameter<int>("camera.width", 1280);
                this->declare_parameter<int>("camera.height", 720);

                // Get parameters
                this->get_parameter("camera.deviceID", deviceID_);
                this->get_parameter("camera.width", width_);
                this->get_parameter("camera.height", height_);

                // Open the camera
                cap_.open(deviceID_);
                if (!cap_.isOpened()) {
                    RCLCPP_ERROR(this->get_logger(), "Failed to open camera");
                    rclcpp::shutdown();
                }

                // Set camera properties
                cap_.set(cv::CAP_PROP_FRAME_WIDTH, width_);
                cap_.set(cv::CAP_PROP_FRAME_HEIGHT, height_);
                std::cout << "Set Frame Width: " << width_ << std::endl;
                std::cout << "Set Frame Height: " << height_ << std::endl;

                cv::Mat frame;
                cap_ >> frame;

                // Current cam roperties
                std::cout << "Current camera properties:" << std::endl;
                std::cout << "Device ID: " << deviceID_ << std::endl;
                std::cout << "Frame Width: " << frame.cols << std::endl;
                std::cout << "Frame Height: " << frame.rows << std::endl;

                // Initialize start time
                start_time_ = std::chrono::steady_clock::now();
            }
            ~CameraNode() {
                if (cap_.isOpened()) {
                    cap_.release();
                    RCLCPP_INFO(this->get_logger(), "Camera resource released.");
                }
            }

        void init() {
            // Initialize ImageTransport and create a publisher
            image_transport::ImageTransport it(shared_from_this());
            pub_ = it.advertise("camera/image_raw", 1);

            // Start the timer for capturing and publishing images
            timer_ = this->create_wall_timer(
                std::chrono::milliseconds(16),
                std::bind(&CameraNode::CaptureAndPublish, this));
        }

    private:
        void CaptureAndPublish() {
            cv::Mat frame;
            cap_ >> frame;
            if (!frame.empty()) {
                // Increment frame count
                frame_count_++;

                // Calculate FPS every second
                std::chrono::time_point<std::chrono::steady_clock> now = std::chrono::steady_clock::now();
                std::chrono::duration<double> elapsed = now - start_time_;
                if (elapsed.count() >= 1.0) {
                    fps_ = frame_count_ / elapsed.count();
                    frame_count_ = 0;
                    start_time_ = now;
                }

                // Convert FPS to string
                std::ostringstream fps_text;
                fps_text << "FPS: " << std::fixed << std::setprecision(2) << fps_;

                // Put FPS text on the frame
                cv::putText(frame, fps_text.str(), cv::Point(10, 30),cv::FONT_HERSHEY_SIMPLEX, 1.0, cv::Scalar(0, 255, 0), 2);

                // Convert OpenCV image to ROS message
                std_msgs::msg::Header header;
                header.stamp = this->now();
                sensor_msgs::msg::Image::SharedPtr msg = cv_bridge::CvImage(header, "bgr8", frame).toImageMsg();

                // Publish the image
                pub_.publish(msg);
            }
        }

    int deviceID_;
    int width_;
    int height_;
    cv::VideoCapture cap_;
    image_transport::Publisher pub_;
    rclcpp::TimerBase::SharedPtr timer_;

    // FPS calculation variables
    std::chrono::steady_clock::time_point start_time_;
    int frame_count_;
    double fps_;
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    std::shared_ptr<CameraNode> node = std::make_shared<CameraNode>();
    node->init();  // Call the initialization function after construction
    rclcpp::spin(node);
    node.reset();
    rclcpp::shutdown();
    return 0;
}

File: ros2_ws/src/s5_py_camera/s5_py_camera/camera.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2
from std_msgs.msg import Header
import time

def find_devices():
    for i in range(5): # Check up to 5 devices
        cap_test = cv2.VideoCapture(i)
        if cap_test.isOpened():
            print(f"Device {i} is available.")
        # Release the VideoCapture object
        cap_test.release()

class CameraNode(Node):
    def __init__(self):
        super().__init__('py_camera_node')
        self.get_logger().info("Python Camera node has been started")

        # Check available camera IDs
        # find_devices()
        
        # Declare parameters with default values
        self.declare_parameter('camera.deviceID', 0)
        self.declare_parameter('camera.width', 1280)
        self.declare_parameter('camera.height', 720)

        # Get parameters
        self.deviceID = self.get_parameter('camera.deviceID').get_parameter_value().integer_value
        self.width = self.get_parameter('camera.width').get_parameter_value().integer_value
        self.height = self.get_parameter('camera.height').get_parameter_value().integer_value

        # Open the camera
        self.cap = cv2.VideoCapture(self.deviceID)
        if not self.cap.isOpened():
            self.get_logger().error('Failed to open camera')
            rclpy.shutdown()
        
        # Set camera properties
        self.cap.set(cv2.CAP_PROP_FRAME_WIDTH, self.width)
        self.cap.set(cv2.CAP_PROP_FRAME_HEIGHT, self.height)
        print(f"Set Frame Width: {self.width}")
        print(f"Set Frame Height: {self.height}")

        ret, frame = self.cap.read()

        # Current cam roperties
        print(f"Current camera properties:")
        print(f"Device ID: {self.deviceID}")
        print(f"Frame Width: {frame.shape[1]}")
        print(f"Frame Height: {frame.shape[0]}")

        # Initialize start time
        self.start_time = time.time()
        self.fps = 0.0

        # Initialize CvBridge and create a publisher
        self.bridge = CvBridge()
        self.publisher = self.create_publisher(Image, 'camera/image_raw', 1)

        # Start the timer for capturing and publishing images
        self.timer = self.create_timer(0.016, self.CaptureAndPublish)
        self.frame_count = 0

    def CaptureAndPublish(self):
        ret, frame = self.cap.read()
        if ret:
            # Increment frame count
            self.frame_count += 1

            # Calculate FPS every second
            elapsed_time = time.time() - self.start_time
            if elapsed_time >= 1.0:
                self.fps = self.frame_count / elapsed_time
                self.frame_count = 0
                self.start_time = time.time()
            
            # Convert FPS to string
            fps_text = f"FPS: {self.fps:.2f}"

            # Put FPS text on the frame
            cv2.putText(frame, fps_text, (10, 30), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (0, 255, 0), 2)

            # Convert OpenCV image to ROS message
            header = Header()
            header.stamp = self.get_clock().now().to_msg()
            image_message = self.bridge.cv2_to_imgmsg(frame, encoding='bgr8')
            image_message.header = header

            # Publish the image
            self.publisher.publish(image_message)

def main(args=None):
    rclpy.init(args=args)
    node = CameraNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.cap.release()
        node.get_logger().info("Camera resource released.")
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()
